lib/api/units/validateBody.js

```diff
@@ -0,0 +1,233 @@
+const jph = require('json-parse-helpfulerror');
+const mediaTyper = require('media-typer');
+const contentTypeUtils = require('content-type');
+
+const { TextDiff } = require('../../../lib/validators/text-diff');
+const { JsonExample } = require('../../../lib/validators/json-example');
+const { JsonSchema } = require('../../../lib/validators/json-schema');
+
+function isPlainText(mediaType) {
+  return mediaType.type === 'text' && mediaType.subtype === 'plain';
+}
+
+function isJson(mediaType) {
+  return mediaType.type === 'application' && mediaType.subtype === 'json';
+}
+
+function isJsonWeak(mediaType) {
+  return isJson(mediaType) || mediaType.suffix === 'json';
+}
+
+/**
+ * Parses a given content-type header into media type.
+ * @param {string} contentType
+ * @returns {[Error, MediaType]}
+ */
+function parseContentType(contentType) {
+  try {
+    const { type } = contentTypeUtils.parse(`${contentType}`);
+    const mediaType = mediaTyper.parse(type);
+    return [null, mediaType];
+  } catch (error) {
+    return [error, null];
+  }
+}
+
+/**
+ * Determines if a given 'Content-Type' header contains JSON.
+ * @param {string} contentType
+ * @returns {boolean}
+ */
+function isJsonContentType(contentType) {
+  // No need to explicitly check for "contentType" existence.
+  // In case given content type is invalid, "parseContentType"
+  // will return an error, which would be coerced to "false".
+  const [error, mediaType] = parseContentType(contentType);
+
+  // Silence an error on purporse because contentType
+  // may contain any kind of rubbish.
+  return error ? false : isJsonWeak(mediaType);
+}
+
+/**
+ * Returns a tuple of error and body media type based
+ * on the given body and normalized headers.
+ * @param {string} body
+ * @param {Object} headers
+ * @returns {[error, bodyType]}
+ */
+function getBodyType(body, headers) {
```

Comment:
This would complicate every place we call `getBodyType` to include safety check:

```js
getBodyType(..., real.headers && real.headers['content-type']
```

However, this can be handled during some sanitization of headers to make sure `headers` is always set to an empty object, when not present. Not sure how it would align with the existing Gavel's behavior, so I wouldn't introduce this change now.

Your suggestion is good, on the other hand.

Reply:
`getBodyType` now accepts content-type header directly: 4a9853626c67eea8f350adb231cb0841f4dbbfdc

Reply:
👍 Yeah, I'd say we should make sure we can rely on the fact headers is always an object.